pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseForwardIndexBasedIndexHandler.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.index.loader;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.SegmentMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreatorProvider;
+import org.apache.pinot.segment.spi.store.ColumnIndexType;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Base class for all of the {@link IndexHandler} classes which need to build their indexes from the forward index.
+ * This class handles temporarily rebuilding the forward index if the forward index does not exist and cleaning up
+ * the forward index once all handlers have completed via overriding the postUpdateIndicesCleanup() method.
+ */
+public abstract class BaseForwardIndexBasedIndexHandler implements IndexHandler {
```

Comment:
I am wondering if this inheritance hierarchy for ForwardIndexHandler actually helps with better abstraction and / or extracting out some common code. 

I am ok with it but it's probably not needed imo. May be once you clarify the confusion on previous comment, I will understand better.

Reply:
Left a comment on the previous comment. Agree that this inheritance hierarchy isn't actually needed but I preferred doing it this way to ensure that we don't duplicate too much code around creating temporary forward index and cleaning it up for all `IndexHandlers` that need the forward index to create their respective indexes. Otherwise all the logic in this class will have to be duplicated across most `IndexHandler` classes.

Reply:
Agreed on extracting out common part of creating and cleaning up temporary forward index. 

Is it possible to fold this into a utility class with public static methods instead ?

Alternatively, will it make sense to call it `BaseIndexHandler` instead with the same purpose of extracting out common piece into it and let concrete implementations override the abstract implementation when needed ?

Another question for future changes - going forward if someone wants to implement the `IndexHandler` interface, then they should extend from `BaseForwardIndexBasedIndexHandler` if the index generation needs fwd index otherwise directly implement `IndexHandler` ? Is that the expectation ?

Reply:
> Is it possible to fold this into a utility class with public static methods instead ?

So I specifically wanted to implement the `postUpdateIndicesCleanup` so that not all `IndexHandler` classes need to implement this. So don't really like the idea of a utility class as I'll need to implement this method to call the utility.

> Alternatively, will it make sense to call it BaseIndexHandler instead with the same purpose of extracting out common piece into it and let concrete implementations override the abstract implementation when needed ?

`BaseIndexHandler` is an option. Do you suggest that all `IndexHandler` classes extend from this, and the classes that don't use the forward index to build their indexes just override the `postUpdateIndicesCleanup` to be a no-op?

> Another question for future changes - going forward if someone wants to implement the IndexHandler interface, then they should extend from BaseForwardIndexBasedIndexHandler if the index generation needs fwd index otherwise directly implement IndexHandler ? Is that the expectation ?

Yes that was indeed the expectation

Reply:
> BaseIndexHandler is an option. Do you suggest that all IndexHandler classes extend from this, and the classes that don't use the forward index to build their indexes just override the postUpdateIndicesCleanup to be a no-op?

Something like that. Btw, I don't meant to be a stickler. So I am fine if you feel strongly about this hierarchy. 

Was just thinking if this fork in the hierarchy is really needed or we can do the typical approach `interface -> abstract class -> all concrete subclass (reuse or override could be no-op too)` type stuff to simplify ?

Reply:
Yeah sure let me try out this suggestion and get back to you if I think it's not a good idea with reasons why. Don't have any objection to trying to simplify this and avoid the fork at the moment.

Reply:
Updated based on your suggestion. let me know what you think

Reply:
Looks good. Thank you